### ISSUE TITLE
[Feature request (re-open)] Python-style triple-quoted strings

### ISSUE THREAD

**Issue by digitalsignalperson** [fish-shell/fish-shell#9694]
I found this related requests from the past: https://github.com/fish-shell/fish-shell/issues/4322

> If anyone makes a good argument for introducing triple-quotes, we can reopen.

A good argument for me is that with triple-quotes I could write and/or copy-paste multi-line strings without worrying about escaping single or double quotes. A specific use-case is using multi-line python code in the shell like this:

```
python -c """
x = 'doing stuff with "quotes" without escaping stuff.'
x += "This is just literal python code copy-pasted from elsewhere"
print(x)
"""
```

This is super convenient to write and/or copy-paste code from anywhere and not need to go in and escape single or double quotes. Maybe it also solves the "here documents" use-case (https://github.com/fish-shell/fish-shell/issues/540).

For further motivation for this python use-case, I'm working on a solution to run `jupyter kernel` as a background process to have a persistent python kernel in my shell, and then have another script that can execute commands in it (alias below is `,python`). Here's an example how I would use the hypothetical triple quotes:

E.g.
```
# Start a kernel in the background
,python --start

# run commands in the background kernel
,python "import os"

# use a magic command
,python "files = !ls | grep .txt"

# run a multiline command with unescaped single and double quotes
,python """
for f in files:
    print(f'Renaming textfile "{f}" to "{f}.old"')
    os.rename(f, f + '.old')
"""
```
etc.

I like the idea of getting to use fish history for jupyter console this way, getting to use ipython magic stuff, all while staying in the terminal and having the flexibility of that.

So in summary, there's an advantage to be able write and/or copy-paste, in this case, python code, without needing to worry about escaping multi-line strings with fish.

**Comment by faho** [comment]
>A good argument for me is that with triple-quotes I could write and/or copy-paste multi-line strings without worrying about escaping single or double quotes. A specific use-case is using multi-line python code in the shell like this:

The pasting use-case mostly goes away once you realize that fish automatically escapes pastes in single-quotes.

You can write

```fish
python -c '
```

and paste your code and it goes through correctly - fish automatically switches the `'` to `\'`.

**Comment by digitalsignalperson** [comment]
Ah I see, that is cool. Yeah that is an argument towards the pasting part, although there are some downsides:
- you can't copy-paste that chunk of code snippet from your history back into a python script or REPL without deleting all the escaping
- cumbersome to paste a chunk and edit in place, possibly with Ctrl+E. You're still wrangling with escaping

I also notice that it works with `python -c '` but not `python -c "`, although I don't think it matters.

Aside from the pasting case, the problem remains for the case of wanting to edit the code directly at the terminal prompt. Most of the time right now I'm writing stuff on the fly in the terminal prompt, and using Alt+E to edit the current command in micro, which is super handy when it gets to multiple lines. It would be an improvement to not have to think about escaping strings.

**Comment by faho** [comment]
>I also notice that it works with python -c ' but not python -c ", although I don't think it matters.

Yup, single-quotes only, like I said. The reason for that is that the escaping is much easier (only backslashes and single-quotes), and that in double-quotes you might want variable interpolation.

Imagine pasting `$HOME/foo`, and it escapes the `$`.

**Comment by ridiculousfish** [comment]
I am in favor of a multi-line string syntax; this keeps coming up. I'm torn between triple-quotes and here documents (#540). HereDocs are objectively disgusting but very widely supported; I lean towards them for that reason.

**Comment by faho** [comment]
> HereDocs are objectively disgusting but very widely supported; I lean towards them for that reason.

I am quite surprised to hear that. I find heredocs absolutely terrible and would hope we never include them.

To go into the issues a bit, as a refresher: "HereDocs" are supplied on stdin, which doesn't actually match the request here.

They have multiple versions - `<<` here "documents", `<<<` here "strings", you can quote and customize the delimiter (and if you quote the delimiter it inhibits expansions on the doc!), if you use `<<-` it'll remove leading *tab* characters (but only tab! not spaces!).

A lot of this is weird and counter-intuitive and rarely useful. It's overly flexible, awkward and makes the `<` character much more complicated.

There are, however, a few useful bits there:

1. Quoting that doesn't clash with the quoted text - you can change the delimiter to not occur in the quoted text
2. Removing leading indentation is useful for embedded python scripts specifically (but restricting it to tabs is terrible)

----

Here's a rough idea:

Quoted blocks. You start a quoted block with at least three of the same quote: `'''` or `"""`. You end it with the same number of the same quote:

```fish
"""""

echo """hahah not ended"""
""""" # <- this ends it
```

So if your quoted text contains the same quotes, simply add more.

Additionally, the first newline is ignored (that block above doesn't start with a `\n`). Probably the final newline as well.

And, here's the kicker: The first line's starting indentation (i.e. run of tabs or spaces) will be removed from every line. I'm not entirely sure if there needs to be a way to turn that *off*, but you could always do:

```fish
    # some fish code above so we're already indented
    somecommand """"
# first line not indented (so nothing is removed!)
    # this line is
""""
```

which is a tad ugly but not worse than what you need to do for python now which is:

https://github.com/fish-shell/fish-shell/blob/df3f2d678ce39ee74f9bfb81d06fcd0ad5a1c598/share/completions/bower.fish#L53-L58

In double-quoted blocks double-quote expansions happen and in single-quoted blocks single-quote expansions (i.e. none) happen. Escaping is as usual - this protects the *quotes*, nothing else (here docs also wouldn't!).

**Comment by digitalsignalperson** [comment]
>  "HereDocs" are supplied on stdin

+1 to that point. Trying to assign a heredoc to a variable or pass as an arg isn't very pretty/intuitive. I'd like the flexibility of using the multiline string as an argument directly.


> So if your quoted text contains the same quotes, simply add more.

I like this.


> Additionally, the first newline is ignored (that block above doesn't start with a \n). Probably the final newline as well.

> The first line's starting indentation (i.e. run of tabs or spaces) will be removed from every line

These make sense to me. It's a nice quality of life improvement where we don't have this in python, except for in function docstrings. In docstrings they use a trim() function as shown here https://peps.python.org/pep-0257/ which I think describes what you are suggesting:

- split the string into lines
- determine minimum indentation (ignoring first line)
- remove the minimum indentation from each line (except the first)
- and then strip off trailing and leading blank lines (but I think we want to limit it to strip only the first and/or last newline, not consecutive ones).
- return all the lines joined by '\n'

I think "determine minimum indentation" makes sense. I don't know if your "The first line's starting indentation" implies a different approach or is the same.

Playing some more with this, I think these would be valid implementations of the example snippet you provided:

min indentation 8
```fish
    if set -l python (__fish_anypython)
        # Warning: That weird indentation is necessary, because python.
        $python -S -c """import json, sys; data = json.load(sys.stdin);
        for k,v in data["dependencies"].items():
            print(k + "\t" + v[:18])""" <bower.json 2>/dev/null
        return
    end
```

min indentation 8
```fish
    if set -l python (__fish_anypython)
        # Warning: That weird indentation is necessary, because python.
        $python -S -c """
        import json, sys
        data = json.load(sys.stdin)
        for k,v in data["dependencies"].items():
            print(k + "\t" + v[:18])
        """ <bower.json 2>/dev/null
        return
    end
```

min indentation 0
```fish
    if set -l python (__fish_anypython)
        # Warning: That weird indentation is necessary, because python.
        $python -S -c """
import json, sys
data = json.load(sys.stdin)
for k,v in data["dependencies"].items():
    print(k + "\t" + v[:18])
""" <bower.json 2>/dev/null
        return
    end
```

**Comment by faho** [comment]
>I think "determine minimum indentation" makes sense. I don't know if your "The first line's starting indentation" implies a different approach or is the same.

It's a different approach. Determining minimum indentation requires scanning the entire string before you can trim.

In particular ignoring the first line complicates the rule a bit, for instance your first example:

```fish
        $python -S -c """import json, sys; data = json.load(sys.stdin);
        for k,v in data["dependencies"].items():
            print(k + "\t" + v[:18])""" <bower.json 2>/dev/null
```

This is supposed to be equivalent to

```fish
        $python -S -c """
        import json, sys; data = json.load(sys.stdin);
        for k,v in data["dependencies"].items():
            print(k + "\t" + v[:18])""" <bower.json 2>/dev/null
```

- because it takes the second line's indentation. This would also be possible without strictly going by the "minimum" indentation, the only real difference would occur in code like

```fish
echo """
     foo
  bar
     baz
     """
```

with minimum indentation this would be "foo" with 3 spaces, "bar" without, "baz" with three spaces and then a line consisting of 3 spaces. To remove that last line it would have to be

```fish
echo """
     foo
  bar
     baz
  """
```

If you picked the first indentation instead, it's not entirely clear how indented that "bar" should be - it could either be an error (bad) or it would remove *up to* the first indentation and it would be unindented.

It would also be impossible to have a first indented line that's more indented than the following - which would be a rare occurence, but a bit of an awkward limitation, so picking the minimum is probably the winner.

> and then strip off trailing and leading blank lines (but I think we want to limit it to strip only the first and/or last newline, not consecutive ones).

Yeah, the idea is to make `"""foo` mean the same as `"""` and then `foo` on the next line. Empty lines at the start/end should still be expressible.

Stripping off the newline is nice so you can place the ending `"""` on a new line without having that weird `print("foo")"""` last line - it allows you to avoid touching the embedded text.

**Comment by piechologist** [comment]
I'd appreciate triple-quoted strings too. It's a feature I regularly miss in fish. IMHO, HereDocs are just ugly. I spend most of my time coding Julia and there's a similar concept of minimum dedentation as discussed above. In the following example, the last line containing the closing triple-quotes determines the dedentation:

```julia
function f()
    println("""
           Hello,
           world.
         """)
end
# prints "  Hello,\n  world.\n"
```

For reference: [Julia Docs, Triple-Quoted-String-Literals](https://docs.julialang.org/en/v1/manual/strings/#Triple-Quoted-String-Literals)

This concept works very well in practice and personally, I'd be thrilled to have the same syntax in Fish and Julia 😎. No more clusters of `echo "..."` in a fish function 🎉.

**Comment by krobelus** [comment]
I have repeatedly stumbled over the lack of heredocs; I think we should add them and perhaps a better quoting style as well.

**Comment by faho** [comment]
I can see adding a thing to pass multiline strings to a command.

But implementing heredocs as they are, with all the garbage, all the associated guff and awkwardness, just seems horrible to me. It's one of the more awkward syntactic corners of bash script.

**Comment by krobelus** [comment]
The common case of `<< EOF` would help to copy-paste single commands from shell scripts

**Comment by faho** [comment]
>The common case of << EOF would help to copy-paste single commands from shell scripts

That goes for literally every bash feature.

It's *not* an extremely common single-line thing, we've covered most of those by now - most of the time when I see some shell instructions in a blog post they'll also work for fish.

I believe this feature is misdesigned and awkward to use, and if we adopted it I would have to question why we're doing this at all. At what point do we simply give up and be posix-compatible?

**Comment by valpackett** [comment]
"Fun" case for triple-quoted strings: when a fish command is templated by some silly program that cannot escape quotes in the substitutions… like a fuzzy-finder like [skim](https://github.com/lotabout/skim) with the `--preview` flag…

Well, even better would be to have raw string literals, like Rust's `r#####"whatever-{can-comfortably-insert-quoted-nonsense-here}"#####` (allows an arbitrary number of `#` that must match on both sides)

**Comment by tmccombs** [comment]
Another use case for this would be something like a json string with embedded variable substitution. 

It's annoying to have to escape all the double quotes. I suppose you could do something like:

```
'{"a": "b", "c": "'$c'"}'
```
but... it's not great. I'd much prefer some form of quotes that allowed both unquoted `"` and variable substitution.

**Comment by Drowze** [comment]
Just wanted to add another use-case that I particularly use a lot: syntax highlight (on the text editor) in a different language for a given block of code.

Neovim have great support for that in languages that support heredoc, as long as the delimiter is named after the embedded language.

For an example, consider a simple sh/bash script:
```bash
# This a simple file with bash syntax highlighting
for i in 1 2 3 4 5; do
  if [ $(( i % 2 )) -eq 0 ]; then
    echo "$i is even"
  else
    echo "$i is odd"
  fi
done

# But it also contains an embedded Ruby script with Ruby syntax highlighting
echo "Ruby script output:"
cat <<'RUBY' | ruby
=begin
This is a multi-line
comment in Ruby
=end
[1, 2, 3].each do |n|
  puts n * 2
end
RUBY

# And a simple Python script with Python syntax highlighting
echo "Python script output:"
cat <<'PYTHON' | python3
str = '''
This is a multi-line
string in Python
'''
for i in range(5):
  print(i ** 2)
PYTHON

# And a simple SQL query with SQL syntax highlighting
echo "SQL query output:"
cat <<'SQL' | sqlite3 :memory:
-- This is a comment in SQL
CREATE TABLE users (id INTEGER PRIMARY KEY, name TEXT);
INSERT INTO users (name) VALUES ('Alice'), ('Bob'), ('Charlie');
SELECT * FROM users;
SQL

# And a simple Lua script with Lua syntax highlighting
cat <<'LUA' | lua
-- This is a comment in Lua
for i = 1, 5 do
  print(i * 3)
end
LUA
```

When editing the above script in neovim, most setups (having at least tree-sitter enabled and the respective grammars installed) would embed a different syntax highlight for the duration of the heredoc, which is very nice feature. For instance, this is how the above look in my neovim:
<img width="666" height="835" alt="Image" src="https://github.com/user-attachments/assets/4557ade2-4173-4f9e-b08a-4b0eaddd5e16" />

Since fish does not support heredocs or anything similar, I'm not aware of any way to do the same in a fish script (and I can't think of any nice way of implementing something similar with custom tree-sitter queries either)

---

To reproduce the neovim case in specific, here's a minimal neovim configuration with tree-sitter enabled along with some parsers (just enough for the above snippet):
```lua
-- NOTE: this requires the nightly (0.12+) version of neovim, as the built-in plugin
-- manager `vim.pack` is not available on earlier versions. You may also need the
-- tree-sitter cli utility to compile and install grammars.
vim.pack.add({ 'nvim-treesitter/nvim-treesitter' })
require('nvim-treesitter').install({ 'bash', 'lua', 'python', 'ruby', 'sql' }):wait()
vim.api.nvim_create_autocmd('FileType', {
  pattern = { 'bash', 'lua', 'python', 'ruby', 'sh', 'sql' },
  callback = function() vim.treesitter.start() end,
})
vim.cmd.colorscheme('evening') -- a built-in colorscheme with good contrast
```

**Comment by visrosa** [comment]
+1 for triple quotes, -1 for heredocs.

I like the idea of triple single-quotes for not expanding variables and triple double-quotes for allowing variable expansion, makes it an intuitive extension of current quoting behavior.

Besides the usecases of snippets of code in other languages and complex quoteful strings,  triple quotes would be very convenient for:

- Printing Usage and Help output in scripts.

```fish
function foo
    if test -z "$argv"
        echo \
        """
        Usage:
            foo [--option1|--option2] bar
        """
        return
    end
end
```

- More readably listing completion options and their descriptions when it doesn't make sense to use a function to generate it:

```fish
set __pkcon_cmds """
    get-groups     List package groups
    get-filters    List filters
    search         Search for packages
    install        Install packages
    install-local  Install local package files
    download       Download packages to directory"""

complete -c pkcon -n "__fish_use_subcommand" -a '$__pkcon_cmds'
```


As for how to deal with indentation, it would be nice to have some control over it but I'm not sure how to not fall into arcane syntax that isn't very fishy. I thought of something like

-  `"""= [...strings...] """` (or maybe `"""+`) to preserve all  indentation and whitespace verbatim

- `"""- [...strings...] """` to trim all extraneous whitespace and newlines (useful to lay out strings more readably in the source code without affecting it) 

- the standard `"""` would use whatever heuristic is most intuitive.

- `"""$ [...strings...] """` this is probably going against the idea of triple quoting as a way to represent a literal verbatim block of text and might be the least fishy of all my ideas, but given the usecase of printing usage information, `--help` texts or more complex errors, this would allow command substitution inside the triple-quoted string and  the use of `set_color` and the like.
